crates/revm/src/state_change.rs

```diff
@@ -51,11 +56,84 @@ pub fn post_block_balance_increments(
     balance_increments
 }
 
-/// Applies the pre-block call to the EIP-4788 beacon block root contract, using the given block,
+/// todo: temporary move over of constants from revm until we've migrated to the latest version
+pub const HISTORY_SERVE_WINDOW: u64 = 8192;
+
+/// Applies the pre-block state change outlined in [EIP-2935] to store historical blockhashes in a
+/// system contract.
+///
+/// If Prague is not activated, or the block is the genesis block, then this is a no-op, and no
+/// state changes are made.
+///
+/// If the provided block is after Prague has been activated, the parent hash will be inserted.
+///
+/// [EIP-2935]: https://eips.ethereum.org/EIPS/eip-2935
+#[inline]
+pub fn apply_blockhashes_update<DB: Database + DatabaseCommit>(
+    db: &mut DB,
+    chain_spec: &ChainSpec,
+    block_timestamp: u64,
+    block_number: u64,
+    parent_block_hash: B256,
+) -> Result<(), BlockExecutionError>
+where
+    DB::Error: std::fmt::Display,
+{
+    // If Prague is not activated or this is the genesis block, no hashes are added.
+    if !chain_spec.is_prague_active_at_timestamp(block_timestamp) || block_number == 0 {
+        return Ok(())
+    }
+    assert!(block_number > 0);
+
+    // Account is expected to exist either in genesis (for tests) or deployed on mainnet or
+    // testnets.
+    // If the account for any reason does not exist, we create it with the EIP-2935 bytecode and a
+    // nonce of 1, so it does not get deleted.
+    let mut account: Account = db
+        .basic(HISTORY_STORAGE_ADDRESS)
+        .map_err(|err| BlockValidationError::Eip2935StateTransition { message: err.to_string() })?
+        .unwrap_or_else(|| AccountInfo {
+            nonce: 1,
+            code: Some(Bytecode::new_raw(HISTORY_STORAGE_CODE.clone())),
+            ..Default::default()
+        })
+        .into();
+
+    // Insert the state change for the slot
+    let (slot, value) = eip2935_block_hash_slot(db, block_number - 1, parent_block_hash)
+        .map_err(|err| BlockValidationError::Eip2935StateTransition { message: err.to_string() })?;
```

Comment:
small nit. Would move `map_err` inside `eip2935_block_hash_slot` so we could have `eip2935_block_hash_slot(..)?`

eip2935_block_hash_slot is specialized helper function.